### DISCUSSION TITLE
guides/model-deployment-practices/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#14818] [Docs]
# guides/model-deployment-practices/

Learn essential tips, insights, and best practices for deploying computer vision models with a focus on efficiency, optimization, troubleshooting, and maintaining security.

https://docs.ultralytics.com/guides/model-deployment-practices/

**Comment by Taufiq-Ahmed** [comment]
Hi, I Have trained my custom dataset on yolov8s model. Now I want to optimized the model by using pruning and quantization. could you please tell me How I can implement it and what is the python code of it?

**Reply by pderrenger** [reply]
@Taufiq-Ahmed to optimize your YOLOv8 model using pruning and quantization, you can follow these steps:

1. **Pruning**: This involves removing weights that contribute little to the final output, making the model smaller and faster. You can use libraries like TensorFlow Model Optimization Toolkit or PyTorch's pruning methods.

2. **Quantization**: This converts the model's weights and activations from high precision (e.g., 32-bit floats) to lower precision (e.g., 8-bit integers), reducing model size and speeding up inference. You can use TensorFlow Lite for post-training quantization or PyTorch's quantization tools.

For detailed guidance and code examples, please refer to the [Ultralytics documentation on model optimization](https://docs.ultralytics.com/guides/model-deployment-practices/). This resource provides comprehensive instructions and best practices for deploying and optimizing your models effectively.

**Reply by Taufiq-Ahmed** [reply]
Thank you so much for your reply. One think I would like to know, is it possible to do pruning and quantization in my YOLOv8 best.pt model with TensorFlow Model Optimization Toolkit or PyTorch's pruning methods. Do do I need to convert it and also need to convert my dataset from YOLO format to any other format?

**Reply by pderrenger** [reply]
Yes, you can perform pruning and quantization on your YOLOv8 model using TensorFlow Model Optimization Toolkit or PyTorch's pruning methods. You don't need to convert your dataset format, but you may need to convert your model to a compatible format like ONNX if using TensorFlow tools. For PyTorch, you can directly apply pruning and quantization techniques. For detailed steps, please refer to the [Ultralytics documentation on model deployment practices](https://docs.ultralytics.com/guides/model-deployment-practices/).

**Reply by Taufiq-Ahmed** [reply]
Thank you for your reply. could you please give me any git hub repositories or any coding example on how to apply pruning and quantization techniques on the best.pt model on custom dataset. and how will i maintain the accuracy of the model.

**Reply by pderrenger** [reply]
@Taufiq-Ahmed for pruning and quantization of your YOLOv8 model, you can use PyTorch's built-in functionalities. You don't need to convert your dataset format. For pruning, you can use `torch.nn.utils.prune`, and for quantization, you can use `torch.quantization`. If you need specific code examples, please let me know. For more detailed guidance, you can refer to the [Ultralytics documentation on model deployment practices](https://docs.ultralytics.com/guides/model-deployment-practices/).

**Reply by Taufiq-Ahmed** [reply]
Thank you for your reply. it would be better for me if you share a coding example of how to prune and quantize the model on the custom data to reduce it's size and weights using pytorch and how to ´maintain it's accuracy. do I need to train it again or there is some other way to maintain it's accuracy.

**Reply by Taufiq-Ahmed** [reply]
Hi could you please check whether this idea works:

import torch
import torch.nn as nn
from torch.nn.utils import prune
from ultralytics import YOLO


model = YOLO('best.pt')


def gradual_prune_model(model, prune_step=0.05, prune_iterations=10):
    for i in range(prune_iterations):
 
        for name, module in model.model.named_modules():
            if isinstance(module, nn.Conv2d):
                prune.l1_unstructured(module, name='weight', amount=prune_step)
                prune.remove(module, 'weight')  # Remove pruning reparameterization

        print(f'Fine-tuning after pruning iteration {i+1}/{prune_iterations}')
        model.train(data='dataset.yaml', epochs=2, lr=1e-4)


gradual_prune_model(model, prune_step=0.05, prune_iterations=10)


print('Final fine-tuning after pruning')
model.train(data='/path/to/your/dataset.yaml', epochs=10, lr=1e-5)

model.save('pruned_yolov8.pt')



is this pruned technique works? and after pruning  is this the right way to regain it's accuracy?

**Reply by pderrenger** [reply]
Your approach to pruning looks good. Gradual pruning followed by fine-tuning can help maintain accuracy. After pruning, fine-tuning with a lower learning rate is essential to regain any lost accuracy. For quantization, you can use PyTorch's quantization tools. Ensure you validate the model's performance after each step. For more detailed guidance, refer to the [Ultralytics documentation on model deployment practices](https://docs.ultralytics.com/guides/model-deployment-practices/).
